nvflare/app_common/workflows/base_model_controller.py

```diff
@@ -379,6 +376,13 @@ def sample_clients(self, num_clients: int = None) -> List[str]:
 
         return clients
 
+    def set_fl_context(self, data: FLModel):
+        """Set up the fl_ctx information based on the passed in FLModel data."""
+        if data and data.current_round is not None:
+            self.fl_ctx.set_prop(AppConstants.CURRENT_ROUND, data.current_round, private=True, sticky=True)
+        else:
+            self.warning("The FLModel data does not contain the current_round information.")
+
```

Comment:
any risk moving from `_process_result` to before @SYangster

Reply:
That was the kludge that we had to do in order to make the FLModel based approach still work with other FLContext/Event based components (such as best model selectors).

In FLModel based approach, both Data and Contextual info (such as current round) are in the FLModel object. But someone must place contextual info into the FLContext, otherwise those event based components will break.

We looked everywhere and just couldn't find a good place to do it. For example, the current round is supposed to be owned by the workflow, but the workflow writer could be a researcher that "we want to hide details from", and hence he won't know things like FLContext.

So what can we do? The next logical place that can put contextual info into FLContext is those "send" methods, which are under our control. There is a requirement that the user must do: set the current round in the FLModel object!

I don't really like this solution, but this is the best we can do right now.

I think we need to take another hard look at this, and come up with something better.